glean-core/rlb/src/private/memory_distribution.rs

```diff
@@ -0,0 +1,147 @@
+use std::sync::Arc;
+
+use glean_core::metrics::{DistributionData, MemoryUnit, MetricType};
+use glean_core::ErrorType;
+
+use crate::dispatcher;
+
+// We need to wrap the glean-core type: otherwise if we try to implement
+// the trait for the metric in `glean_core::metrics` we hit error[E0117]:
+// only traits defined in the current crate can be implemented for arbitrary
+// types.
+
+/// This implements the developer-facing API for recording memory distribution metrics.
+///
+/// Instances of this class type are automatically generated by the parsers
+/// at build time, allowing developers to record values that were previously
+/// registered in the metrics.yaml file.
+#[derive(Clone)]
+pub struct MemoryDistributionMetric(pub(crate) Arc<glean_core::metrics::MemoryDistributionMetric>);
+
+impl MemoryDistributionMetric {
+    /// The public constructor used by automatically generated metrics.
+    pub fn new(meta: glean_core::CommonMetricData, memory_unit: MemoryUnit) -> Self {
+        Self(Arc::new(
+            glean_core::metrics::MemoryDistributionMetric::new(meta, memory_unit),
+        ))
+    }
+}
+
+#[inherent(pub)]
+impl glean_core::traits::MemoryDistribution for MemoryDistributionMetric {
+    /// Accumulates the provided sample in the metric.
+    ///
+    /// # Arguments
+    ///
+    /// * `sample` - The sample to be recorded by the metric. The sample is assumed to be in the
+    ///   configured memory unit of the metric.
+    ///
+    /// ## Notes
+    ///
+    /// Values bigger than 1 Terabyte (2<sup>40</sup> bytes) are truncated
+    /// and an `ErrorType::InvalidValue` error is recorded.
+    fn accumulate(&self, sample: u64) {
+        let metric = Arc::clone(&self.0);
+        dispatcher::launch(move || crate::with_glean(|glean| metric.accumulate(glean, sample)));
+    }
+
+    /// Accumulates the provided signed samples in the metric.
+    ///
+    /// This is required so that the platform-specific code can provide us with
+    /// 64 bit signed integers if no `u64` comparable type is available. This
+    /// will take care of filtering and reporting errors for any provided negative
+    /// sample.
+    ///
+    /// Please note that this assumes that the provided samples are already in
+    /// the "unit" declared by the instance of the implementing metric type
+    /// (e.g. if the implementing class is a [MemoryDistribution] and the
+    /// instance this method was called on is using kilobyte, then `samples` are
+    /// assumed to be in that unit).
+    ///
+    /// # Arguments
+    ///
+    /// * `samples` - The vector holding the samples to be recorded by the metric.
+    ///
+    /// ## Notes
+    ///
+    /// Discards any negative value in `samples` and report an `ErrorType::InvalidValue`
+    /// for each of them.
+    /// Values bigger than 1 Terabyte (2<sup>40</sup> bytes) are truncated
+    /// and an `ErrorType::InvalidValue` error is recorded.
+    fn accumulate_samples_signed(&self, samples: Vec<i64>) {
```

Comment:
I wonder if we need this. We only really needed it to power the FFI layer for Kotlin. Maybe we should remove it from the bindings as well. @badboy , thoughts?

Reply:
Happy to remove it from the trait as part of this work if so.

Reply:
Hm. By that logic, should we also remove `test_get_value_as_json_string`? Rust consumers don't need/want it, to my knowledge.

Reply:
> 
> 
> Hm. By that logic, should we also remove `test_get_value_as_json_string`? Rust consumers don't need/want it, to my knowledge.

Yes, I believe we want to remove that as well, but let's see what @badboy thinks.

Reply:
Yup, let's do the minimal viable traits and expand them when we require new APIs